WordPress/gutenberg
# Interactivity API: initial support for SSR

## What?



This PR adds server-side rendering for the following directives:

- wp-context
- wp-bind
- wp-class
- wp-style
- wp-text

Tracking issue: https://github.com/WordPress/gutenberg/issues/51056

## Why?



Server-side rendering of the directives ensures that the HTML that the server sends will match the first HTML generated after the initial hydration improving the UX and SEO.

## How?



Most of the code was created originally by @ockham in the [block-interactivity-experiments repo](https://github.com/WordPress/block-hydration-experiments) with the help of @dmsnell and others, and ported by me to Gutenberg.

The current implementation uses either the HTML Tag Processor or a new class built on top of it (WP Directive Processor) that adds things like `set_inner_html` which are necessary for `wp-text`.

**IMPORTANT DISCLAIMER: This code of the WP_Directive_Processor class is highly experimental and its only purpose is to provide a way to test the server-side rendering of the Interactivity API. Most of this code will be discarded once the HTML Processor is available. Please restrain from investing unnecessary time and effort trying to improve this code.**

The technique to process only the root nodes is new (it was not in the BIE repo) but it doesn't yet work as it should because it processes the root blocks of each post loaded (the template, the template parts, the post content, the reusable blocks…), when it should only process the root blocks of the template to ensure that directives are only processed once. Therefore, there are still multiple processing of the same directives happening. Still, I would prefer to merge this PR and try to fix it in a subsequent one.

## Testing Instructions






- Open a post
- Add an HTML block
- Paste this code
  ```html
  <div data-wp-context='{ "hide": false, "color": "red" }'>
    <div
      class="another"
      style="text-decoration: underline"
      data-wp-bind--hidden="context.hide"
      data-wp-class--show="!context.hide"
      data-wp-style--color="context.color"
      data-wp-text="context.color"
    >
      default value
    </div>
  </div>
  ```
- Take a look at the frontend, there should be a red underlined text saying "red" with the class "show"
- Change the "color" to "blue".
- Now there should be a blue underlined text saying "blue"
- Change "hide" to true.
- The text now should have a `hidden` attribute and the "show" class should not be there anymore
